plugins/filter_cgolib/cgolib.c

```diff
@@ -0,0 +1,276 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2015-2018 Treasure Data Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_filter.h>
+#include <fluent-bit/flb_utils.h>
+#include <fluent-bit/flb_pack.h>
+#include <fluent-bit/flb_parser.h>
+#include <fluent-bit/flb_unescape.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#ifdef _WIN32
+#include <fluent-bit/flb_dlfcn_win32.h>
+#else
+#include <dlfcn.h>
+#endif
+
+#include <msgpack.h>
+#include "cgo.h"
+
+#include "cgolib.h"
+
+// copy from filter_parser.c
+static int msgpackobj2char(msgpack_object *obj,
+                           char **ret_char, int *ret_char_size)
+{
+    int ret = -1;
+
+    if (obj->type == MSGPACK_OBJECT_STR) {
+        *ret_char      = (char*)obj->via.str.ptr;
+        *ret_char_size = obj->via.str.size;
+        ret = 0;
+    }
+    else if (obj->type == MSGPACK_OBJECT_BIN) {
+        *ret_char      = (char*)obj->via.bin.ptr;
+        *ret_char_size = obj->via.bin.size;
+        ret = 0;
+    }
+
+    return ret;
+}
+
+/*
+ * unpack msg, call filter , write result to parameter result
+ */
+static int callCgoLibFilter(struct cgolib_conf *ctx, msgpack_sbuffer *result, char *tag, int tag_len, void *data, size_t bytes)
+{
+    int index= 0, i;
+    char *key_str, *val_str;
+    int key_len, val_len;
+
+    msgpack_object_kv *kv;
+    size_t off = 0;
+    msgpack_object *obj;
+    int map_num;
+    struct flb_time tm;
+
+    GoSlice name_slice, value_slice;
+    _GoString_ name[MAX_FIELD]={0}, value[MAX_FIELD]={0};
```

Comment:
Should we use `GoString` instead of raw `_GoString_` type?

Reply:
This is create by go  build, I suggest do not change it.

Reply:
Got it. Thanks. :smile: